lib/engine/game/g_22_mars/game.rb

```diff
@@ -168,18 +168,105 @@ class Game < Game::Base
           },
         ].freeze
 
+        # Game specific constants
         LAST_OR = 11
+        FIRST_REVOLT_CHECK_OR = 7
+        LAST_REVOLT_CHECK_OR = 10
 
-        @or = 0
+        def setup
+          @or = 0
+          @three_or_round = false
+          @revolt_round = determine_revolt_round
+        end
 
         def end_now?(_after)
           @or == LAST_OR
         end
 
+        def setup_preround
+          revolts, permits = @companies.partition(&:revolt?)
+          sorted_permits = permits.sort_by { rand }
+          sorted_revolts = revolts.sort_by { rand }
+
+          @companies = sorted_permits.take(@players.size)
+          assigned_companies = sorted_permits.drop(@players.size) + sorted_revolts
+
+          assigned_companies.zip(@corporations).each do |company, corporation|
+            if corporation
+              company.owner = corporation
+              corporation.companies << company
+              @log << "#{company.name} is assigned to #{corporation.name}"
+            else
+              company.close!
+              @log << "#{company.name} is discarded"
+            end
+          end
+        end
+
+        def unowned_purchasable_companies
+          @companies
+        end
+
+        def init_round
+          G22Mars::Round::PermitPurchase.new(self, [G22Mars::Step::SelectOrDiscard])
+        end
+
         def new_operating_round(round_num = 1)
           @or += 1
 
-          super
+          if @or == 9
+            @operating_rounds = 3
+            @three_or_round = true
+          end
+
+          round = super
+
+          check_revolt!
+
+          round
+        end
+
+        def or_round_finished
+          # In case we get phase change during the last OR set we ensure we have 3 ORs
+          @operating_rounds = 3 if @three_or_round
+        end
+
+        def revolt_happened?
+          @revolt_happened
+        end
+
+        def determine_revolt_round
+          (FIRST_REVOLT_CHECK_OR..LAST_REVOLT_CHECK_OR).each do |round|
+            probability = LAST_REVOLT_CHECK_OR - round + 1
+
+            return round if (rand % probability).zero?
```

Comment:
This will work, but a simpler version would be:
```suggestion
          (FIRST_REVOLT_CHECK_OR..LAST_REVOLT_CHECK_OR).to_a.sort_by { rand }.first
```

Reply:
Thanks for the tip, it should be identical. I'll try it out in the next part